src/mame/sega/sega_beena.cpp

```diff
@@ -71,131 +102,2074 @@
+	    in memory to find its exact start address, then consume enough cycles to
+	    reduce the number of instructions executed until the next IRQ is asserted.
+	*/
+	if (m_irq_wait_start_addr == UNKNOWN_ADDR) {
+		if (m_maincpu->pc() > 0xc00cc000 && m_maincpu->pc() < 0xc00fff80) {
+			int8_t addr_delta = 8;
+			memory_share *shared = machine().root_device().memshare("workram");
+			uint32_t *shared32 = reinterpret_cast<uint32_t *>(shared->ptr());
+			uint32_t candidate_start_addr = m_maincpu->pc() - addr_delta;
+			uint32_t candidate_offset = (candidate_start_addr - 0xc00cc000) / 4;
+			for (size_t i = 0; i < addr_delta; i++) {
+				bool matched = true;
+				for (size_t sig_i = 0; sig_i < 5; sig_i++) {
+					if (irq_wait_signature[sig_i] != shared32[candidate_offset + i + sig_i]) {
+						matched = false;
+						break;
+					}
+				}
+				if (matched) {
+					m_irq_wait_start_addr = candidate_start_addr + i;
+
+					for (size_t sig_i = 0; sig_i < 5; sig_i++) {
+						m_maincpu->add_hotspot(candidate_start_addr + i + sig_i * 4);
+					}
+				}
+			}
+		}
+	}
+}
+
+static const gfx_layout sega_beena_8bpp_layout =
+{
+	16,16,
+	0x800, // 0x800 * 0x10 * 0x10 = 0x80000 (tile data mapping size)
+	8,
+	{ STEP8(0, 1) },
+	{ STEP16(0, 8) },
+	{ STEP16(0, 8*16) },
+	8*16*16
+};
+
+void sega_9h0_0008_state::beena_arm7_map(address_map &map)
+{
+	// BIOS internal ROM
+	map(0x00000000, 0x0001ffff).rom();
+	// BIOS internal RAM
+	map(0x20000000, 0x20003fff).ram();
+
+	// FIXME: Need to confirm upper bounds / mirrored ranges on hardware, some return inconsistent reads
+	// Video (registers, palette, sprite data...)
+	map(0x40000000, 0x400000ff).rw(FUNC(sega_9h0_0008_state::video_reg_r), FUNC(sega_9h0_0008_state::video_reg_w)).mirror(0xff00);
+	map(0x40010000, 0x400103ff).ram().share("tilemap_sprites");
+	map(0x40010400, 0x4001ffff).ram();
+	map(0x40020000, 0x400201ff).w(FUNC(sega_9h0_0008_state::pal_w)).share("paletteram");
+	map(0x40020200, 0x4002ffff).ram();
+	map(0x40030000, 0x400302ff).ram().share("tilemap_scroll_y");
+	map(0x40030300, 0x4003ffff).ram();
+
+	// I/O (registers, pages, pads, pen...)
+	map(0x50000000, 0x5000000f).ram().mirror(0xfff0); // Unknown
+	map(0x50010000, 0x5001003f).lrw32(
+		NAME([this](offs_t offset) {
+			LOG("audio_pcm_reg_r @ %08x: addr=%08x data=%08x\n", m_maincpu->pc(), offset * 4, m_pcm->reg_r(offset));
+
+			return m_pcm->reg_r(offset);
+		}),
+		NAME([this](offs_t offset, uint32_t data, uint32_t mem_mask) {
+			LOG("audio_pcm_reg_w @ %08x: addr=%08x data=%08x mask=%08x\n", m_maincpu->pc(), offset * 4, data, mem_mask);
+
+			// FIXME: Games hang during .au data processing unless FIQ requests are served,
+			// review this after playback is implemented
+			if (offset == 0x10/4) {
+				if (data != 0) {
+					m_requested_fiq = true;
+				}
+			}
+
+			m_pcm->reg_w(offset, data, mem_mask);
+		})).mirror(0xffc0);
+	map(0x50020000, 0x5002007f).r(FUNC(sega_9h0_0008_state::io_sensors_r)).share("io_sensor_regs").mirror(0xff80);
```

Comment:
For all the entries, please put the mirror value before the handlers, so it’s next to the addresses.  It makes interpreting the memory map easier if all the addressing stuff is together.